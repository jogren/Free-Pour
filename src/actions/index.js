export const showSelectDrink = targetDrink => ({
  type: 'SHOW_SELECT_DRINK',
  targetDrink
});

export const hideSelectedDrink = () => ({
  type: 'HIDE_SELECTED_DRINK'
});

export const setCurrentCocktails = cocktails => ({
  type: 'SET_CURRENT_COCKTAILS',
  cocktails
});