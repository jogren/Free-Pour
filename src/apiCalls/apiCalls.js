export const fetchPopularCocktails = async () => {
  const url = 'https://www.thecocktaildb.com/api/json/v2/8673533/popular.php';
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error('There was an error fetch your data')
    }
    const popularCocktails = await response.json()
    return popularCocktails.drinks.map(drink => ({
      strDrink: drink.strDrink,
      strDrinkThumb: drink.strDrinkThumb,
    }))
  } catch(error) {
    throw new Error(error.message)
  }
}

export const fetchCocktailsByGenre = async (type) => {
  const url = `https://www.thecocktaildb.com/api/json/v1/1/filter.php?i=${type}`
  try {
    const response = await fetch(url)
    if(!response.ok) {
      throw new Error('There was an issue fetching your data by genre')
    }
    const cocktails = await response.json();
    return cocktails.drinks
  } catch (error) {
    throw new Error(error.message)

  }
}

export const fetchCocktailsBySearch = async (text) => {
  const url = `https://www.thecocktaildb.com/api/json/v1/1/search.php?s=${text}`
  try {
    const response = await fetch(url)
    if(!response.ok) {
      throw new Error('There was an issue fetching your data by name')
    }
    const cocktails = await response.json();
    return cocktails.drinks
  } catch(error) {
    throw new Error(error.message)
  }
}

export const fetchAllIngredients = async () => {
  const url = '  https://www.thecocktaildb.com/api/json/v1/1/list.php?i=list';
  try {
    const response = await fetch(url)
    if (!response.ok) {
      throw new Error('There was an issue fetching all ingredients')
    }
    const ingredients = await response.json();
    return ingredients.drinks.map(drink => drink.strIngredient1)
  } catch(error) {
    throw new Error(error.message)
  }
}

export const fetchMoreDrinkInfo = async (name) => {
  try {
    const response = await fetch(`https://www.thecocktaildb.com/api/json/v1/1/search.php?s=${name.split(' ').join('_')}`)
    if (!response.ok) {
      throw new Error('There was an error fetch your data')
    }
    const cocktail = await response.json();
    const cleanedCocktail = cocktail.drinks.map(drink => ({
      name: drink.strDrink,
      glass: drink.strGlass,
      instructions: drink.strInstructions,
      image: drink.strDrinkThumb,
      ingredients: [
        { measure: drink.strMeasure1, ingredient: drink.strIngredient1 },
        { measure: drink.strMeasure2, ingredient: drink.strIngredient2 },
        { measure: drink.strMeasure3, ingredient: drink.strIngredient3 },
        { measure: drink.strMeasure4, ingredient: drink.strIngredient4 },
        { measure: drink.strMeasure5, ingredient: drink.strIngredient5 },
        { measure: drink.strMeasure6, ingredient: drink.strIngredient6 }
      ]
    }))
    return cleanedCocktail[0]
  } catch(error) {
    throw new Error(error.message)
  }
}