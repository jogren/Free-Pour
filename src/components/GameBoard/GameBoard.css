.GameBoard_main {
  display: flex;
  flex-direction: column;
  align-items: center;
  position: relative;
}

.GameBoard_main h3 {
  margin: 25px;
  font-size: 24px;
  text-align: center;
}

.GameBoard_input {
  height: 30px;
  border: 2px solid black;
  font-size: 12px;
  font-weight: 700;
  border-radius: 5px 0px 0px 5px;
  margin-top: 2px;
  padding-left: 5px;
  width: 200px;
}

.GameBoard_input::placeholder { 
  color: black;
}

.button-home {
  position: absolute;
  top: 10px;
  right: 10px;
}

.button-home:hover {
  background-color: black;
  color: #FFF;
}

.wrong-answer-error {
  position: absolute;
}