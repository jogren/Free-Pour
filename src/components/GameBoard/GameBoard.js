import React, { Component } from 'react';
import GameBoardContainer from '../GameBoardContainer/GameBoardContainer';
import { connect } from 'react-redux';
import { fetchAllIngredients } from '../../apiCalls/apiCalls';
import { toggleFavorite } from '../../actions';
import { NavLink } from 'react-router-dom';
import wrongAnswerImage from '../../images/wrong-answer.png';
import './GameBoard.css';

class GameBoard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      roundCounter: 0,
      ingredientSearch: '',
      allIngredients: [],
      roundFinished: false,
      wrongAnswerError: ''
    }
  }

  componentDidMount = async () => {
    try {
      const allIngredients = await fetchAllIngredients();
      this.setState({ allIngredients: allIngredients })
    } catch (error) {
      throw new Error(error.message)
    }
  }

  handleSubmitGuess = () => {
    let { roundCounter, ingredientSearch } = this.state;
    const { favoriteCocktails } = this.props;
    let isCorrect = favoriteCocktails[roundCounter].ingredients.find(ingredient => ingredient.ingredient.toLowerCase().includes(ingredientSearch.toLowerCase()))
    if(isCorrect) {
      isCorrect.guessed = true
    } else {
      this.setState({ wrongAnswerError: 'wrong' })
      setTimeout(() => { this.setState({ wrongAnswerError: '' }) }, 500);
    }
    let allTrue = favoriteCocktails[roundCounter].ingredients.every(ingredient => ingredient.guessed || ingredient.ingredient === '')
    if (allTrue && roundCounter + 1 >= favoriteCocktails.length) {
      this.state.roundFinished = true
    } else if (allTrue) {
      this.state.roundCounter++
    }
    this.setState({ ingredientSearch: '' })
  }

  handleIngredientSearch = e => {
    this.setState({ [e.target.name]: e.target.value })
  }

  resetGame = () => {
    let { favoriteCocktails, toggleFavorite } = this.props;
    let favoriteCocktailsReset = favoriteCocktails.map(cocktail => {
        return {
          name: cocktail.name,
          glass: cocktail.glass,
          image: cocktail.image,
          instructions: cocktail.instructions,
          ingredients: cocktail.ingredients.map(ingredient => {
            return {
              measure: ingredient.measure,
              ingredient: ingredient.ingredient,
              guessed: false
            }
          })
        }
      })
    toggleFavorite(favoriteCocktailsReset)
    this.setState({ roundCounter: 0, roundFinished: false })
  }

  render() {
    const { favoriteCocktails } = this.props;
    const { roundCounter, ingredientSearch, allIngredients } = this.state;
    let addMoreIngredientOptions = [...allIngredients, 'Coca-Cola', 'Olive', 'Soda Water', 'Cherry', 'Mint', 'Blue Curacao', 'Angostura Bitters']
    let ingredientList = addMoreIngredientOptions.map((ingredient, index) => {
      return <option key={index} value={ingredient}/>
    })
    return (
      <main className="GameBoard_main">
        <h3>What's in a {favoriteCocktails[roundCounter].name}?</h3>
        <GameBoardContainer currentDrink={favoriteCocktails[this.state.roundCounter]}/>
        <section className="Board_section-ingredients">
          <input 
            type="text"
            placeholder="Search for Ingredients..."
            name="ingredientSearch"
            value={ingredientSearch}
            onChange={(e) => this.handleIngredientSearch(e)}
            list="ingredient-list"
          />
          <datalist id="ingredient-list">{ingredientList}</datalist>
          <button onClick={this.handleSubmitGuess}>Submit Guess</button>
        </section>
        {this.state.roundFinished && (
          <div>
            <h3>Congrats! You reviewed all of your favorites</h3> 
            <NavLink to="/">
              <button onClick={this.resetGame}>Keep studying!</button>
            </NavLink>
          </div>
        )}
        {this.state.wrongAnswerError && <img src={wrongAnswerImage}/> }
      </main>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  toggleFavorite: (cocktails) => dispatch(toggleFavorite(cocktails))
})

export default connect(null, mapDispatchToProps)(GameBoard);

