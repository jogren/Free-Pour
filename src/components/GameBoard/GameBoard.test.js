import React from 'react';
import { shallow } from 'enzyme';
import { GameBoard, mapDispatchToProps } from './GameBoard';
import { fetchAllIngredients } from '../../apiCalls/apiCalls';
import { toggleFavorite } from '../../actions';

jest.mock('../../apiCalls/apiCalls');

describe('GameBoard', () => {
  let wrapper;
  const mockToggleFavorite = jest.fn();
  const mockFavoriteCocktails = [
    {
      name: 'Moscow Mules',
      glass: 'Copper Mug',
      instructions: 'Combine ingredients...',
      image: 'image2.png',
      ingredients: [
        {
          guessed: false,
          ingredient: 'Vodka',
          measure: '2-3 oz'
        },
        {
          guessed: false,
          ingredient: 'Ginger Ale',
          measure: '5oz'
        },
        {
          guessed: false,
          ingredient: 'Lime Juice',
          measure: '2tbsp'
        }
      ]
    }
  ]

  beforeEach(() => {
    wrapper = shallow(<GameBoard 
      favoriteCocktails={mockFavoriteCocktails}
      toggleFavorite={mockToggleFavorite}
      />)
    })
    
    it('should take a snapshot when passed the correct data', () => {
      expect(wrapper).toMatchSnapshot();
    });
    
    it('should invoke handleIngredientSearch onChange of input', () => {
      wrapper.instance().handleIngredientSearch = jest.fn();
      wrapper.instance().forceUpdate();
      
      wrapper.find('input').simulate('change')
      expect(wrapper.instance().handleIngredientSearch).toHaveBeenCalled();
    })
    
  describe('componentDidMount', () => {
    it('should call fetchAllIngredients', () => {
      expect(fetchAllIngredients).toHaveBeenCalled();
    })
  })

  describe('resetGame', () => {
    it('should reset the game and structure cocktails', () => {
      const expected = [
        {
          name: 'Moscow Mules',
          glass: 'Copper Mug',
          instructions: 'Combine ingredients...',
          image: 'image2.png',
          ingredients: [
            {
              guessed: false,
              ingredient: 'Vodka',
              measure: '2-3 oz'
            },
            {
              guessed: false,
              ingredient: 'Ginger Ale',
              measure: '5oz'
            },
            {
              guessed: false,
              ingredient: 'Lime Juice',
              measure: '2tbsp'
            }
          ]
        }
      ]
      wrapper.instance().resetGame();
      expect(mockToggleFavorite).toHaveBeenCalledWith(expected)
      expect(wrapper.state('roundCounter')).toEqual(0)
      expect(wrapper.state('roundFinished')).toEqual(false)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with an toggleFavorite action', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = toggleFavorite([{ id: 1, name: 'Moscow Mule' }, { id: 2, name: 'Vodka Soda' }]);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.toggleFavorite([{ id: 1, name: 'Moscow Mule' }, { id: 2, name: 'Vodka Soda' }])
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
})