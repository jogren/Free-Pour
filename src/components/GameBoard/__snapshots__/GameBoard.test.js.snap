// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard should take a snapshot when passed the correct data 1`] = `
<main
  className="GameBoard_main"
>
  <h3>
    What's in a 
    Moscow Mules
    ?
  </h3>
  <GameBoardContainer
    currentDrink={
      Object {
        "glass": "Copper Mug",
        "image": "image2.png",
        "ingredients": Array [
          Object {
            "guessed": false,
            "ingredient": "Vodka",
            "measure": "2-3 oz",
          },
          Object {
            "guessed": false,
            "ingredient": "Ginger Ale",
            "measure": "5oz",
          },
          Object {
            "guessed": false,
            "ingredient": "Lime Juice",
            "measure": "2tbsp",
          },
        ],
        "instructions": "Combine ingredients...",
        "name": "Moscow Mules",
      }
    }
  />
  <section
    className="Board_section-ingredients"
  >
    <input
      list="ingredient-list"
      name="ingredientSearch"
      onChange={[Function]}
      placeholder="Search for Ingredients..."
      type="text"
      value=""
    />
    <datalist
      id="ingredient-list"
    >
      <option
        key="0"
        value="Coca-Cola"
      />
      <option
        key="1"
        value="Olive"
      />
      <option
        key="2"
        value="Soda Water"
      />
      <option
        key="3"
        value="Cherry"
      />
      <option
        key="4"
        value="Mint"
      />
      <option
        key="5"
        value="Blue Curacao"
      />
      <option
        key="6"
        value="Angostura Bitters"
      />
    </datalist>
    <button
      onClick={[Function]}
    >
      Submit Guess
    </button>
  </section>
</main>
`;
